File: app/services/feedback_service.py
# app/services/feedback_service.py

"""
Service layer for customer feedback and service closure workflow
"""

from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, desc
from fastapi import HTTPException, status
from datetime import datetime
import json
import logging

from app.models import (
    CustomerFeedback, ServiceClosure, InstallationJob, CompletionRecord,
    Customer, User, Organization
)
from app.schemas.feedback import (
    CustomerFeedbackCreate, CustomerFeedbackUpdate, CustomerFeedbackFilter,
    ServiceClosureCreate, ServiceClosureUpdate, ServiceClosureFilter,
    FeedbackStatus, ClosureStatus, ClosureReason
)

logger = logging.getLogger(__name__)


class CustomerFeedbackService:
    """Service for customer feedback operations"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_feedback(
        self, 
        feedback_data: CustomerFeedbackCreate, 
        organization_id: int
    ) -> CustomerFeedback:
        """Create a new customer feedback record"""
        
        # Validate installation job exists and belongs to organization
        installation_job = self.db.query(InstallationJob).filter(
            and_(
                InstallationJob.id == feedback_data.installation_job_id,
                InstallationJob.organization_id == organization_id
            )
        ).first()
        
        if not installation_job:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Installation job not found"
            )
        
        # Validate customer belongs to organization
        customer = self.db.query(Customer).filter(
            and_(
                Customer.id == feedback_data.customer_id,
                Customer.organization_id == organization_id
            )
        ).first()
        
        if not customer:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Customer not found"
            )
        
        # Check if feedback already exists for this job
        existing_feedback = self.db.query(CustomerFeedback).filter(
            and_(
                CustomerFeedback.installation_job_id == feedback_data.installation_job_id,
                CustomerFeedback.customer_id == feedback_data.customer_id,
                CustomerFeedback.organization_id == organization_id
            )
        ).first()
        
        if existing_feedback:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Feedback already exists for this job"
            )
        
        # Serialize survey responses to JSON if provided
        survey_responses_json = None
        if feedback_data.survey_responses:
            survey_responses_json = json.dumps(feedback_data.survey_responses)
        
        # Create feedback record
        feedback = CustomerFeedback(
            organization_id=organization_id,
            installation_job_id=feedback_data.installation_job_id,
            customer_id=feedback_data.customer_id,
            completion_record_id=feedback_data.completion_record_id,
            overall_rating=feedback_data.overall_rating,
            service_quality_rating=feedback_data.service_quality_rating,
            technician_rating=feedback_data.technician_rating,
            timeliness_rating=feedback_data.timeliness_rating,
            communication_rating=feedback_data.communication_rating,
            feedback_comments=feedback_data.feedback_comments,
            improvement_suggestions=feedback_data.improvement_suggestions,
            survey_responses=survey_responses_json,
            would_recommend=feedback_data.would_recommend,
            satisfaction_level=feedback_data.satisfaction_level,
            follow_up_preferred=feedback_data.follow_up_preferred,
            preferred_contact_method=feedback_data.preferred_contact_method,
            feedback_status=FeedbackStatus.SUBMITTED
        )
        
        self.db.add(feedback)
        self.db.commit()
        self.db.refresh(feedback)
        
        # Update completion record feedback status if linked
        if feedback_data.completion_record_id:
            completion_record = self.db.query(CompletionRecord).filter(
                CompletionRecord.id == feedback_data.completion_record_id
            ).first()
            if completion_record:
                completion_record.feedback_request_sent = True
                completion_record.feedback_request_sent_at = datetime.utcnow()
                self.db.commit()
        
        logger.info(f"Created customer feedback {feedback.id} for job {feedback_data.installation_job_id}")
        return feedback
    
    def get_feedback_by_id(self, feedback_id: int, organization_id: int) -> Optional[CustomerFeedback]:
        """Get feedback by ID"""
        return self.db.query(CustomerFeedback).filter(
            and_(
                CustomerFeedback.id == feedback_id,
                CustomerFeedback.organization_id == organization_id
            )
        ).first()
    
    def get_feedback_list(
        self, 
        organization_id: int, 
        filter_params: CustomerFeedbackFilter,
        skip: int = 0,
        limit: int = 100
    ) -> List[CustomerFeedback]:
        """Get list of feedback with filters"""
        
        query = self.db.query(CustomerFeedback).filter(
            CustomerFeedback.organization_id == organization_id
        )
        
        # Apply filters
        if filter_params.feedback_status:
            query = query.filter(CustomerFeedback.feedback_status == filter_params.feedback_status)
        
        if filter_params.overall_rating:
            query = query.filter(CustomerFeedback.overall_rating == filter_params.overall_rating)
        
        if filter_params.customer_id:
            query = query.filter(CustomerFeedback.customer_id == filter_params.customer_id)
        
        if filter_params.installation_job_id:
            query = query.filter(CustomerFeedback.installation_job_id == filter_params.installation_job_id)
        
        if filter_params.satisfaction_level:
            query = query.filter(CustomerFeedback.satisfaction_level == filter_params.satisfaction_level)
        
        if filter_params.from_date:
            query = query.filter(CustomerFeedback.submitted_at >= filter_params.from_date)
        
        if filter_params.to_date:
            query = query.filter(CustomerFeedback.submitted_at <= filter_params.to_date)
        
        return query.order_by(desc(CustomerFeedback.submitted_at)).offset(skip).limit(limit).all()
    
    def update_feedback(
        self, 
        feedback_id: int, 
        feedback_update: CustomerFeedbackUpdate, 
        organization_id: int,
        updated_by_id: Optional[int] = None
    ) -> CustomerFeedback:
        """Update feedback record"""
        
        feedback = self.get_feedback_by_id(feedback_id, organization_id)
        if not feedback:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Feedback not found"
            )
        
        # Update fields
        update_data = feedback_update.model_dump(exclude_unset=True)
        
        # Handle survey responses serialization
        if 'survey_responses' in update_data and update_data['survey_responses']:
            update_data['survey_responses'] = json.dumps(update_data['survey_responses'])
        
        for field, value in update_data.items():
            setattr(feedback, field, value)
        
        # Set review information if status is being updated
        if feedback_update.feedback_status and updated_by_id:
            feedback.reviewed_by_id = updated_by_id
            feedback.reviewed_at = datetime.utcnow()
        
        self.db.commit()
        self.db.refresh(feedback)
        
        logger.info(f"Updated customer feedback {feedback_id}")
        return feedback


class ServiceClosureService:
    """Service for service closure workflow operations"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def create_closure(
        self, 
        closure_data: ServiceClosureCreate, 
        organization_id: int,
        created_by_id: int
    ) -> ServiceClosure:
        """Create a new service closure record"""
        
        # Validate installation job exists and belongs to organization
        installation_job = self.db.query(InstallationJob).filter(
            and_(
                InstallationJob.id == closure_data.installation_job_id,
                InstallationJob.organization_id == organization_id
            )
        ).first()
        
        if not installation_job:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Installation job not found"
            )
        
        # Check if closure already exists for this job
        existing_closure = self.db.query(ServiceClosure).filter(
            and_(
                ServiceClosure.installation_job_id == closure_data.installation_job_id,
                ServiceClosure.organization_id == organization_id
            )
        ).first()
        
        if existing_closure:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Service closure already exists for this job"
            )
        
        # Check if completion record exists
        completion_record = None
        if closure_data.completion_record_id:
            completion_record = self.db.query(CompletionRecord).filter(
                CompletionRecord.id == closure_data.completion_record_id
            ).first()
        
        # Check if feedback exists and calculate metrics
        feedback_received = False
        minimum_rating_met = False
        if closure_data.customer_feedback_id:
            feedback = self.db.query(CustomerFeedback).filter(
                CustomerFeedback.id == closure_data.customer_feedback_id
            ).first()
            if feedback:
                feedback_received = True
                minimum_rating_met = feedback.overall_rating >= 3  # Configurable threshold
        
        # Create closure record
        closure = ServiceClosure(
            organization_id=organization_id,
            installation_job_id=closure_data.installation_job_id,
            completion_record_id=closure_data.completion_record_id,
            customer_feedback_id=closure_data.customer_feedback_id,
            closure_status=ClosureStatus.PENDING,
            closure_reason=closure_data.closure_reason,
            closure_notes=closure_data.closure_notes,
            requires_manager_approval=closure_data.requires_manager_approval,
            approval_notes=closure_data.approval_notes,
            final_closure_notes=closure_data.final_closure_notes,
            feedback_received=feedback_received,
            minimum_rating_met=minimum_rating_met,
            escalation_required=closure_data.escalation_required,
            escalation_reason=closure_data.escalation_reason,
            reopening_reason=closure_data.reopening_reason
        )
        
        self.db.add(closure)
        self.db.commit()
        self.db.refresh(closure)
        
        logger.info(f"Created service closure {closure.id} for job {closure_data.installation_job_id}")
        return closure
    
    def get_closure_by_id(self, closure_id: int, organization_id: int) -> Optional[ServiceClosure]:
        """Get closure by ID"""
        return self.db.query(ServiceClosure).filter(
            and_(
                ServiceClosure.id == closure_id,
                ServiceClosure.organization_id == organization_id
            )
        ).first()
    
    def get_closure_list(
        self, 
        organization_id: int, 
        filter_params: ServiceClosureFilter,
        skip: int = 0,
        limit: int = 100
    ) -> List[ServiceClosure]:
        """Get list of closures with filters"""
        
        query = self.db.query(ServiceClosure).filter(
            ServiceClosure.organization_id == organization_id
        )
        
        # Apply filters
        if filter_params.closure_status:
            query = query.filter(ServiceClosure.closure_status == filter_params.closure_status)
        
        if filter_params.closure_reason:
            query = query.filter(ServiceClosure.closure_reason == filter_params.closure_reason)
        
        if filter_params.requires_manager_approval is not None:
            query = query.filter(ServiceClosure.requires_manager_approval == filter_params.requires_manager_approval)
        
        if filter_params.feedback_received is not None:
            query = query.filter(ServiceClosure.feedback_received == filter_params.feedback_received)
        
        if filter_params.escalation_required is not None:
            query = query.filter(ServiceClosure.escalation_required == filter_params.escalation_required)
        
        if filter_params.approved_by_id:
            query = query.filter(ServiceClosure.approved_by_id == filter_params.approved_by_id)
        
        if filter_params.closed_by_id:
            query = query.filter(ServiceClosure.closed_by_id == filter_params.closed_by_id)
        
        if filter_params.from_date:
            query = query.filter(ServiceClosure.created_at >= filter_params.from_date)
        
        if filter_params.to_date:
            query = query.filter(ServiceClosure.created_at <= filter_params.to_date)
        
        return query.order_by(desc(ServiceClosure.created_at)).offset(skip).limit(limit).all()
    
    def approve_closure(
        self, 
        closure_id: int, 
        organization_id: int,
        approved_by_id: int,
        approval_notes: Optional[str] = None
    ) -> ServiceClosure:
        """Approve a service closure (manager action)"""
        
        closure = self.get_closure_by_id(closure_id, organization_id)
        if not closure:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Service closure not found"
            )
        
        if closure.closure_status != ClosureStatus.PENDING:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Only pending closures can be approved"
            )
        
        closure.closure_status = ClosureStatus.APPROVED
        closure.approved_by_id = approved_by_id
        closure.approved_at = datetime.utcnow()
        if approval_notes:
            closure.approval_notes = approval_notes
        
        self.db.commit()
        self.db.refresh(closure)
        
        logger.info(f"Approved service closure {closure_id} by user {approved_by_id}")
        return closure
    
    def close_service(
        self, 
        closure_id: int, 
        organization_id: int,
        closed_by_id: int,
        final_closure_notes: Optional[str] = None
    ) -> ServiceClosure:
        """Finalize service closure (manager action)"""
        
        closure = self.get_closure_by_id(closure_id, organization_id)
        if not closure:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Service closure not found"
            )
        
        if closure.closure_status not in [ClosureStatus.PENDING, ClosureStatus.APPROVED]:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Only pending or approved closures can be closed"
            )
        
        if closure.requires_manager_approval and not closure.approved_by_id:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Manager approval required before closing"
            )
        
        closure.closure_status = ClosureStatus.CLOSED
        closure.closed_by_id = closed_by_id
        closure.closed_at = datetime.utcnow()
        if final_closure_notes:
            closure.final_closure_notes = final_closure_notes
        
        self.db.commit()
        self.db.refresh(closure)
        
        logger.info(f"Closed service {closure_id} by user {closed_by_id}")
        return closure
    
    def reopen_service(
        self, 
        closure_id: int, 
        organization_id: int,
        reopened_by_id: int,
        reopening_reason: str
    ) -> ServiceClosure:
        """Reopen a closed service"""
        
        closure = self.get_closure_by_id(closure_id, organization_id)
        if not closure:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Service closure not found"
            )
        
        if closure.closure_status != ClosureStatus.CLOSED:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Only closed services can be reopened"
            )
        
        closure.closure_status = ClosureStatus.REOPENED
        closure.reopened_count += 1
        closure.last_reopened_at = datetime.utcnow()
        closure.last_reopened_by_id = reopened_by_id
        closure.reopening_reason = reopening_reason
        
        # Reset closure fields
        closure.closed_by_id = None
        closure.closed_at = None
        
        self.db.commit()
        self.db.refresh(closure)
        
        logger.info(f"Reopened service {closure_id} by user {reopened_by_id}")
        return closure

File: frontend/src/services/feedbackService.ts
// Frontend service for customer feedback and service closure workflow API

import api from '../lib/api';

export interface CustomerFeedback {
  id: number;
  organization_id: number;
  installation_job_id: number;
  customer_id: number;
  completion_record_id?: number;
  overall_rating: number;
  service_quality_rating?: number;
  technician_rating?: number;
  timeliness_rating?: number;
  communication_rating?: number;
  feedback_comments?: string;
  improvement_suggestions?: string;
  survey_responses?: any;
  would_recommend?: boolean;
  satisfaction_level?: string;
  follow_up_preferred: boolean;
  preferred_contact_method?: string;
  feedback_status: string;
  reviewed_by_id?: number;
  reviewed_at?: string;
  response_notes?: string;
  submitted_at: string;
  updated_at?: string;
}

export interface ServiceClosure {
  id: number;
  organization_id: number;
  installation_job_id: number;
  completion_record_id?: number;
  customer_feedback_id?: number;
  closure_status: string;
  closure_reason?: string;
  closure_notes?: string;
  requires_manager_approval: boolean;
  approved_by_id?: number;
  approved_at?: string;
  approval_notes?: string;
  closed_by_id?: number;
  closed_at?: string;
  final_closure_notes?: string;
  feedback_received: boolean;
  minimum_rating_met: boolean;
  escalation_required: boolean;
  escalation_reason?: string;
  reopened_count: number;
  last_reopened_at?: string;
  last_reopened_by_id?: number;
  reopening_reason?: string;
  created_at: string;
  updated_at?: string;
}

export interface CustomerFeedbackCreate {
  installation_job_id: number;
  customer_id: number;
  completion_record_id?: number;
  overall_rating: number;
  service_quality_rating?: number;
  technician_rating?: number;
  timeliness_rating?: number;
  communication_rating?: number;
  feedback_comments?: string;
  improvement_suggestions?: string;
  survey_responses?: any;
  would_recommend?: boolean;
  satisfaction_level?: string;
  follow_up_preferred?: boolean;
  preferred_contact_method?: string;
}

export interface ServiceClosureCreate {
  installation_job_id: number;
  completion_record_id?: number;
  customer_feedback_id?: number;
  closure_reason?: string;
  closure_notes?: string;
  requires_manager_approval?: boolean;
  approval_notes?: string;
  final_closure_notes?: string;
  escalation_required?: boolean;
  escalation_reason?: string;
  reopening_reason?: string;
}

export interface FeedbackFilters {
  feedback_status?: string;
  overall_rating?: number;
  customer_id?: number;
  installation_job_id?: number;
  from_date?: string;
  to_date?: string;
  satisfaction_level?: string;
}

export interface ClosureFilters {
  closure_status?: string;
  closure_reason?: string;
  requires_manager_approval?: boolean;
  feedback_received?: boolean;
  escalation_required?: boolean;
  approved_by_id?: number;
  closed_by_id?: number;
  from_date?: string;
  to_date?: string;
}

export interface FeedbackAnalytics {
  period_days: number;
  total_feedback: number;
  average_rating: number;
  positive_feedback: number;
  negative_feedback: number;
  satisfaction_rate: number;
}

export interface ClosureAnalytics {
  period_days: number;
  total_closures: number;
  completed_closures: number;
  escalated_closures: number;
  average_reopens: number;
  completion_rate: number;
}

class FeedbackService {
  private readonly baseUrl = '/api/v1/feedback';

  // Customer Feedback Methods
  async submitFeedback(feedbackData: CustomerFeedbackCreate): Promise<CustomerFeedback> {
    const response = await api.post(`${this.baseUrl}/feedback`, feedbackData);
    return response.data;
  }

  async getFeedbackList(
    filters: FeedbackFilters = {},
    skip: number = 0,
    limit: number = 100
  ): Promise<CustomerFeedback[]> {
    const params = new URLSearchParams({
      skip: skip.toString(),
      limit: limit.toString(),
      ...Object.entries(filters).reduce((acc, [key, value]) => {
        if (value !== undefined && value !== '') {
          acc[key] = value.toString();
        }
        return acc;
      }, {} as Record<string, string>)
    });

    const response = await api.get(`${this.baseUrl}/feedback?${params}`);
    return response.data;
  }

  async getFeedbackById(feedbackId: number): Promise<CustomerFeedback> {
    const response = await api.get(`${this.baseUrl}/feedback/${feedbackId}`);
    return response.data;
  }

  async updateFeedback(
    feedbackId: number,
    updateData: Partial<CustomerFeedbackCreate>
  ): Promise<CustomerFeedback> {
    const response = await api.put(`${this.baseUrl}/feedback/${feedbackId}`, updateData);
    return response.data;
  }

  async reviewFeedback(
    feedbackId: number,
    responseNotes: string
  ): Promise<CustomerFeedback> {
    const response = await api.post(`${this.baseUrl}/feedback/${feedbackId}/review`, {
      response_notes: responseNotes
    });
    return response.data;
  }

  // Service Closure Methods
  async createServiceClosure(closureData: ServiceClosureCreate): Promise<ServiceClosure> {
    const response = await api.post(`${this.baseUrl}/service-closure`, closureData);
    return response.data;
  }

  async getServiceClosures(
    filters: ClosureFilters = {},
    skip: number = 0,
    limit: number = 100
  ): Promise<ServiceClosure[]> {
    const params = new URLSearchParams({
      skip: skip.toString(),
      limit: limit.toString(),
      ...Object.entries(filters).reduce((acc, [key, value]) => {
        if (value !== undefined && value !== '') {
          acc[key] = value.toString();
        }
        return acc;
      }, {} as Record<string, string>)
    });

    const response = await api.get(`${this.baseUrl}/service-closure?${params}`);
    return response.data;
  }

  async getServiceClosureById(closureId: number): Promise<ServiceClosure> {
    const response = await api.get(`${this.baseUrl}/service-closure/${closureId}`);
    return response.data;
  }

  async approveServiceClosure(
    closureId: number,
    approvalNotes?: string
  ): Promise<ServiceClosure> {
    const response = await api.post(`${this.baseUrl}/service-closure/${closureId}/approve`, {
      approval_notes: approvalNotes
    });
    return response.data;
  }

  async closeServiceTicket(
    closureId: number,
    finalClosureNotes?: string
  ): Promise<ServiceClosure> {
    const response = await api.post(`${this.baseUrl}/service-closure/${closureId}/close`, {
      final_closure_notes: finalClosureNotes
    });
    return response.data;
  }

  async reopenServiceTicket(
    closureId: number,
    reopeningReason: string
  ): Promise<ServiceClosure> {
    const response = await api.post(`${this.baseUrl}/service-closure/${closureId}/reopen`, {
      reopening_reason: reopeningReason
    });
    return response.data;
  }

  // Analytics Methods
  async getFeedbackAnalytics(days: number = 30): Promise<FeedbackAnalytics> {
    const response = await api.get(`${this.baseUrl}/feedback/analytics/summary?days=${days}`);
    return response.data;
  }

  async getClosureAnalytics(days: number = 30): Promise<ClosureAnalytics> {
    const response = await api.get(`${this.baseUrl}/service-closure/analytics/summary?days=${days}`);
    return response.data;
  }

  // Utility Methods
  async getFeedbackByJobId(jobId: number): Promise<CustomerFeedback | null> {
    try {
      const feedbackList = await this.getFeedbackList({ installation_job_id: jobId });
      return feedbackList.length > 0 ? feedbackList[0] : null;
    } catch (error) {
      console.error('Error fetching feedback by job ID:', error);
      return null;
    }
  }

  async getClosureByJobId(jobId: number): Promise<ServiceClosure | null> {
    try {
      const closureList = await this.getServiceClosures({ installation_job_id: jobId });
      return closureList.length > 0 ? closureList[0] : null;
    } catch (error) {
      console.error('Error fetching closure by job ID:', error);
      return null;
    }
  }

  // Workflow Integration Methods
  async triggerFeedbackRequest(jobId: number): Promise<void> {
    // This would trigger an automated feedback request
    // Could be called from the dispatch management system
    const response = await api.post(`${this.baseUrl}/feedback/trigger-request`, {
      installation_job_id: jobId
    });
    return response.data;
  }

  async canClosureBeCreated(jobId: number): Promise<boolean> {
    // Check if a service can be closed based on business rules
    try {
      const feedback = await this.getFeedbackByJobId(jobId);
      const existingClosure = await this.getClosureByJobId(jobId);
      
      // Business rules: feedback received and no existing closure
      return feedback !== null && existingClosure === null;
    } catch (error) {
      console.error('Error checking closure eligibility:', error);
      return false;
    }
  }
}

// Create singleton instance
export const feedbackService = new FeedbackService();

// Export types and service
export {
  FeedbackService,
  feedbackService as default
};